File: api/explainFormula.js
const ALLOWED_ORIGINS = new Set([
  'https://www.careersolutionsfortoday.com',
  'https://careersolutionsfortoday.com',
  'https://stevenmkay.github.io',
  'http://localhost:5500',
  'http://127.0.0.1:5500'
]);

export default async function handler(req, res) {
  try {
    applyCors(req, res);
  } catch (err) {
    console.error('CORS configuration error (formula endpoint):', err);
    res.status(500).json({ success: false, error: 'Server configuration error' });
    return;
  }

  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }

  if (req.method !== 'POST') {
    res.status(405).json({ success: false, error: 'Method not allowed' });
    return;
  }

  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    res.status(500).json({ success: false, error: 'Server missing OpenAI credentials' });
    return;
  }

  try {
    const { formula } = req.body || {};
    if (!formula || typeof formula !== 'string' || formula.trim().length < 3) {
      res.status(400).json({ success: false, error: 'A valid formula is required.' });
      return;
    }

    const trimmedFormula = formula.trim();
    const userContent = `Explain the following Excel formula in structured steps.

Formula:
${trimmedFormula}
`;

    const requestBody = {
      model: 'gpt-4.1-mini',
      messages: [
        {
          role: 'system',
          content: 'You are an Excel tutor that explains formulas in friendly, plain English. Always return a short summary, a step-by-step breakdown, and at least one practical usage tip.'
        },
        {
          role: 'user',
          content: userContent
        }
      ],
      temperature: 0.2,
      max_tokens: 800
    };

    const response = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`
      },
      body: JSON.stringify(requestBody)
    });

    if (!response.ok) {
      const errorPayload = await response.json().catch(() => ({}));
      throw new Error(errorPayload?.error?.message || response.statusText || 'OpenAI request failed');
    }

    const data = await response.json();
    const explanation = data?.choices?.[0]?.message?.content?.trim();

    if (!explanation) {
      res.status(502).json({ success: false, error: 'No explanation returned from model.' });
      return;
    }

    res.status(200).json({ success: true, explanation });
  } catch (error) {
    console.error('Excel explainer error:', error);
    res.status(500).json({ success: false, error: error.message || 'Failed to explain formula' });
  }
}

function applyCors(req, res) {
  const origin = req.headers.origin;
  if (origin && (ALLOWED_ORIGINS.has(origin) || process.env.ALLOW_ALL_ORIGINS === 'true')) {
    res.setHeader('Access-Control-Allow-Origin', origin);
  } else {
    res.setHeader('Access-Control-Allow-Origin', 'https://www.careersolutionsfortoday.com');
  }
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
  res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
}

File: api/explain-formula.py
import os
import json
from openai import OpenAI
from http.server import BaseHTTPRequestHandler

# Initialize OpenAI client
client = OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))

# Allowed frontend origins
ALLOWED_ORIGINS = {
    "https://www.careersolutionsfortoday.com",
    "https://careersolutionsfortoday.com",
    "https://stevenmkay.github.io",
    "http://localhost:5500",
    "http://127.0.0.1:5500"
}

def set_cors_headers(handler, origin):
    """Attach proper CORS headers."""
    if origin in ALLOWED_ORIGINS:
        handler.send_header("Access-Control-Allow-Origin", origin)
    else:
        # Safe fallback
        handler.send_header("Access-Control-Allow-Origin", "https://www.careersolutionsfortoday.com")

    handler.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
    handler.send_header("Access-Control-Allow-Headers", "Content-Type")


def safe_json(response):
    """Safely parse JSON without crashing on invalid responses."""
    try:
        return response.json()
    except Exception:
        return None


class handler(BaseHTTPRequestHandler):

    def do_OPTIONS(self):
        origin = self.headers.get("Origin")
        self.send_response(200)
        set_cors_headers(self, origin)
        self.end_headers()

    def do_POST(self):
        origin = self.headers.get("Origin")

        try:
            # --- Read client request ---
            content_length = int(self.headers.get("Content-Length", 0))
            raw_body = self.rfile.read(content_length).decode("utf-8")

            try:
                data = json.loads(raw_body)
            except json.JSONDecodeError:
                return self.respond_json(400, origin, {
                    "success": False,
                    "error": "Invalid JSON in request body."
                })

            formula = data.get("formula", "").strip()

            # --- Validation ---
            if not formula or len(formula) < 3:
                return self.respond_json(400, origin, {
                    "success": False,
                    "error": "A valid Excel formula is required."
                })

            if len(formula) > 2000:
                return self.respond_json(400, origin, {
                    "success": False,
                    "error": "Formula is too long to analyze."
                })

            # Ensure formula starts with "=" but don't double-add
            if not formula.startswith("="):
                formula = "=" + formula

            # --- Build prompts ---
            system_prompt = (
                "You are an Excel tutor that explains formulas in friendly, plain English.\n"
                "Format output using:\n"
                "- Section titles followed by a colon\n"
                "- Numbered main points (1. 2. 3.)\n"
                "- Indented sub-points using exactly 3 leading spaces + '-'\n"
                "- No markdown formatting, no bold, no special symbols\n\n"
                "Indented example:\n"
                "1. First step\n"
                "   - Sub explanation\n"
                "   - More detail\n"
            )

            user_prompt = (
                f"Explain this Excel formula:\n\n{formula}\n\n"
                "Provide the following sections:\n"
                "1. What It Does\n"
                "2. How It Works (use indented sub-bullets)\n"
                "3. Example with sample data (indented sub-bullets)\n"
                "4. Tips (indented sub-bullets)\n\n"
                "NO markdown. Use only plain text with clean indentation."
            )

            # --- NEW RESPONSES API CALL ---
            openai_response = client.responses.create(
                model="gpt-4.1-mini",
                input=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_prompt}
                ],
                temperature=0.3,
                max_output_tokens=1000
            )

            # Extract text safely
            try:
                explanation = (
                    openai_response.output[0]
                    .content[0]
                    .text
                    .strip()
                )
            except Exception:
                return self.respond_json(502, origin, {
                    "success": False,
                    "error": "OpenAI returned an unexpected response format."
                })

            # --- Return successful response ---
            return self.respond_json(200, origin, {
                "success": True,
                "explanation": explanation
            })

        except Exception as e:
            # Your server logs receive the real error
            print("SERVER ERROR:", str(e))

            # The client receives a safe message
            return self.respond_json(500, origin, {
                "success": False,
                "error": "Internal server error. Please try again later."
            })

    def respond_json(self, status, origin, payload):
        """Utility for sending JSON responses with proper CORS."""
        self.send_response(status)
        set_cors_headers(self, origin)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps(payload).encode())
